src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java

```diff
@@ -53,18 +59,48 @@ public static void checkUser(Request req) {
         try {
             DecodedJWT jwt = verifier.verify(token);
             Auth0UserProfile profile = new Auth0UserProfile(jwt);
+            if (!isValidUser(profile)) {
+                // If not a valid user (and not creating an OTP/API user for self), halt the request. Note: creating an
+                // admin user requires that the requester is an admin (checkUserIsAdmin must be passed), so this is not
+                // a concern for that method/controller.
+                if (!isCreatingSelf(req, profile)) {
+                    logMessageAndHalt(req, HttpStatus.NOT_FOUND_404, "Unknown user.");
+                }
+            }
             // The user attribute is used on the server side to check user permissions and does not have all of the
             // fields that the raw Auth0 profile string does.
             addUserToRequest(req, profile);
         } catch (JWTVerificationException e) {
             // Invalid signature/claims
             logMessageAndHalt(req, 401, "Login failed to verify with our authorization provider.", e);
+        } catch (HaltException e) {
+            throw e;
         } catch (Exception e) {
             LOG.warn("Login failed to verify with our authorization provider.", e);
             logMessageAndHalt(req, 401, "Could not verify user's token");
         }
     }
 
+    /**
+     * Check for POST requests that are creating an {@link AbstractUser} (a proxy for OTP/API users).
+     */
+    private static boolean isCreatingSelf(Request req, Auth0UserProfile profile) {
+        // First, check the request method/path.
+        String uri = req.uri();
+        String method = req.requestMethod();
+        if (method.equalsIgnoreCase("POST") && (uri.endsWith(OTP_USER_PATH) || uri.endsWith(API_USER_PATH))) {
+            // Next, get the user object from the request body, verifying that the Auth0UserId matches.
+            try {
+                AbstractUser user = getPOJOFromRequestBody(req, AbstractUser.class);
```

Comment:
A piece of code might still be missing here.
For a new signup from MOD UI, an exception is thrown:
```log
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.opentripplanner.middleware.models.AbstractUser` (no Creators, like default construct, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
 at [Source: (String)"{"auth0UserId":"auth0|5f****************",
<req.body() content below>}"; line: 1, column: 1]
```

`req.body()` returns something such as 
```json
{
"auth0UserId":"auth0|5f**************",
"email":"email@example.com",
"hasConsentedToTerms":true,
"notificationChannel":"email",
"phoneNumber":"",
"savedLocations":[],
"storeTripHistory":false
}
```

Reply:
The easiest workaround is probably to pass `ApiUser.class` or `OtpUser.class` to the get POJO method depending on the path.

Reply:
OK, fixed and tested. https://github.com/ibi-group/otp-middleware/pull/64/commits/869713da7fde4b06d71eb64ac001530ae7f4cc73